### DISCUSSION TITLE
How to use Element-Plus only, forget inefficiency DIV+CSS?

### DISCUSSION THREAD

**Discussion by gimjin** [element-plus/element-plus#11493] [Ideas]
Most developers who use Element-Plus are applicable to information applications (for example: ERP, CMS, more CRUD...)

I see that although Element-Plus is powerful enough, there are still many people who can't use Element-Plus and Theme well.

I tried to use only Element-Plus + Vue-charts + Unocss in my company project, and I was pleasantly surprised to find that almost all div span css (Vue style tag) were destroyed, and the efficiency was greatly improved.

### However, there are still a few places where span css(Vue style tag) cannot be avoided
- Many slots need nested spans to define special colors with unocss
- Similar to el-space, the internal el-space-item cannot use unocss, so it can only be controlled by adding div/span unocss to the outside
- There is an isolated design (such as a line of text), the component is heavy, or only div span can be used
Because of such a small flaw, I can't express it clearly to other developers "Element-Plus only, forget inefficiency DIV+CSS"

### solution
Develop el-block, which is simply a span, which may have optional attributes of div and span

#### Improved
- only Element-Plus + unocss clean and tidy in vue tempalte
- There will be no style and scope style everywhere to make the theme gradually invalid, only use the theme
   - I wrote a plugin [unplugin-remove-vue-style](https://github.com/gimjin/unplugin-remove-vue-style) for company projects, prohibiting other developers from writing private styles in the project
- html semantics, high readability
- GUI Design the boundary between "can develop" and "can' not develop", div+css is too low-level and can do everything

**All nodes of vue template use element-plus components. Unify the style with element-plus theme (only modify the style variable in :root)
Complete upgrade  from div+css to Element-Plus + Theme**

# TODO list
- [x] #11653 
- [x] #11976 
- [ ] #12084

### Note (2023-03-16)

After cleaned up span and div, removed `1374` lines of code(div/span/css) replaced it with unocss.

```bash
# code line statistics
find element-plus/docs/examples -name "*.vue" | xargs cat | grep -v ^$ | wc -l
```
According to the #12084 code statistics, examples folder are total of `18075` lines of code, and `1374` lines (7.6%) have been deleted. And  very time-consuming for developers in non-english-speaking countries to HTML element set classname.

So upgrading from `MIXED div + css` to `ONLY element-plus components + theme` can reduce the workload by at `least 10%`.

**Comment by ryuhangyeong** [comment]
Sounds like a good idea worth discussing!

**Reply by gimjin** [reply]
Thank U :)

I have participated in Nokia Maemo OS Framework App and Android Framework App Theme change, they wrote very good maemo and android framework app are using component library, so rewrite gtkrc and resource xml theme file.
More  UIKit  developer mostly use components, few use of draw view.

Maybe use div+css is very easy or powerful, so...

**Reply by ryuhangyeong** [reply]
Is the component you want something like below?

## Example

- https://mantine.dev/core/box/
- https://mantine.dev/core/paper/
- https://vue.chakra-ui.com/box
- https://vue.chakra-ui.com/text

**Reply by gimjin** [reply]
Exactly, like c-text
```diff
<el-table-column>
+  <el-text class="mt-8px" size="small" type="info">text</el-text>
-   <span class="mt-8px">text</span>
+  <el-text class="mt-8px" tag="p">a paragraph of text...</el-text>
-   <p class="mt-8px">a paragraph of text...</p>
<el-table-column>
```

But maybe the el-box(Box and c-box) is not a good idea, because it can be flooded like div. It's too bad to turn into el-box+css.

**Comment by gimjin** [comment]
Hi @ryuhangyeong 

I developed el-text and created  PR #11653, Waiting for approval, or ignoring some Element-Plus PR rules?

because for the 1st contribution, I need you help :)

**Reply by ryuhangyeong** [reply]
@gimjin 

Thank you for your contribution!

However, adding new components requires empathy from more people. For example, there could be a discussion about the direction or design of an API.

I fully sympathize with the need for the component you suggest, but we need to discuss that need further.

And for the time being, it seems that the addition of new APIs will not be scheduled.

#6944

And I think a great opportunity to incorporate new ideas is when upgrading to a major version.

Please understand with a generous heart!

@jw-foss do you currently accept new features?

**Reply by gimjin** [reply]
Thank you.

**Reply by jw-foss** [reply]
@gimjin actually I was thinking of this component, I think we can accept this one after we adjust the PR to align with out design guidelines  /cc @ryuhangyeong 👍

**Comment by gimjin** [comment]
I will actively develop to align design guidelines merged PR.

**Comment by gimjin** [comment]
Thank you @jw-foss @ryuhangyeong for giving me the opportunity to develop el-text and realize this idea.

I hope to optimize the div and span in the example, this work may change a lot of files, so apply in advance.

Official examples are the best reference to guide many engineers, and I hope that almost all examples like Naive-ui almost all use with own components.

Sincerely hope it will be approved.

### e.g.
```diff
<el-statistic :value="138">
  <template #title>
-    <div style="display: inline-flex; align-items: center">
-      Ratio of men to women
-      <el-icon style="margin-left: 4px" :size="12">
-        <Male />
-      </el-icon>
-    </div>
+    <el-text>
+      Ratio of men to women
+      <el-icon style="margin-left: 4px" :size="12">
+        <Male />
+      </el-icon>
+    </el-text>
  </template>
  <template #suffix>/100</template>
</el-statistic>
```
```diff
-<div class="m-4">
-  <p>Child options expand when clicked (default)</p>
-  <el-cascader v-model="value" :options="options" @change="handleChange" />
-</div>
+<el-row class="m-4">
+  <el-text tag="p">Child oel-text tag="p"tions expand when clicked (default)</el-text>
+  <el-cascader v-model="value" :options="options" @change="handleChange" />
+</el-row>
```
and more...

**Reply by ryuhangyeong** [reply]
@gimjin good job!

**Reply by gimjin** [reply]
Hi @ryuhangyeong I created a PR #12084  (283 changed files with 2,520 additions and 3,893 deletions).
Sorry review is very hard, I passed the self-test. 🙏

**Reply by ryuhangyeong** [reply]
@gimjin to many changes 😂😂
